feat: add status dots for pods

cdrage opened the pull request:
feat: add status dots for pods

### What does this PR do?

* Adds dots for the containers when listing a pod (podman or kubernetes)
* Hover to see the tooltip of the container name as well as the running
  status
* Dots will change when the container status has changed.
* Fixes a small redirection error bug when clicking a kubernetes pod

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast
explaining what is doing this PR -->

https://github.com/containers/podman-desktop/assets/6422176/7fbee0ef-9a2a-42e7-9087-7ab08c29b425




### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes https://github.com/containers/podman-desktop/issues/3925
Closes https://github.com/containers/podman-desktop/issues/4612

### How to test this PR?

<!-- Please explain steps to reproduce -->

cdrage commented:
Ready for review! 💯 

Need some help however with regards to what colours to use as there are 9 different types of statuses that can be shown.

https://github.com/containers/podman-desktop/pull/4646/files#diff-122a4449594e476ae1551ff015e3547ecec1c30d999c77295b5938a329d495f7R17

deboer-tim commented on packages/renderer/src/lib/ui/StatusDot.svelte:
```diff
@@ -0,0 +1,35 @@
+
+  // Podman-only
+  stopped: 'bg-gray-500', // Gray indicated dormancy / inactivity
+  paused: 'bg-amber-500', // Equivilant to "waiting" in Kubernetes
```
-600. and Equivalent :)

benoitf replied:
what about adding a `status` color in the palette (tailwind.config.js) with something like (colors codes are pure example) 

```js
'status': {
        'running': tailwindColors.green[500],
        'stopped': tailwindColors.amber[500],
        etc..
      }
```

so when we need to display a running state we do `bg-status-running` and we see values only at one location

benoitf replied:
or an utility class that provides/share the value across StatusDot and StatusIcon

deboer-tim replied:
+1 to tailwind cfg, I think we could use these for things like machine status too.

deboer-tim commented on packages/renderer/src/lib/ui/StatusDot.svelte:
```diff
@@ -0,0 +1,35 @@
+  exited: 'bg-red-300', // Light red for exited, but dark red for dead
+  dead: 'bg-red-500', // Dark red for dead
+  created: 'bg-green-300', // Blue indicates "freshness", so would be good for 'created'
+  degraded: 'bg-amber-700', // Dark amber indicating "degrading" state / "unhealthy"
```
Should be -600

deboer-tim commented on packages/renderer/src/lib/ui/StatusDot.svelte:
```diff
@@ -0,0 +1,35 @@
+// All the possible statuses that will appear for both Pods and Kubernetes
+const colors: Record<string, string> = {
+  // Podman & Kubernetes
+  running: 'bg-green-500', // Green for running
```
The only existing colours we have are in StatusIcon, so we should start be matching those, e.g. -400 here (or arguing that we should change both). It doesn't have all these states though, so picking reasonable ones for things that aren't covered should be ok.

jeffmaury reviewed (changes requested):
We should handle large numbers of containers better 

![dots](https://github.com/containers/podman-desktop/assets/695993/c91b1b0a-f3e1-4817-a98a-b8eb8d66466d)

cdrage commented:
As a result of some last minute discussions, I'll be implementing the following in the PR:

- [ ] Fix misrepresented colours (some are -700 which should be -400)
- [ ] Add colours representing statuses to the tailwind configuration so we can change the colours there rather than in the svelte file
- [ ] For more than 10 containers, have them separate by colour dot and a number beside them, example: 🔴 +4, with the tooltip saying: "Terminated: 4".
- [ ] The containers should be arranged by terminated / running, for example: running first, then paused / degraded / waiting then terminated / stopped / etc.

cdrage commented:
Ready for another review / test! @jeffmaury @benoitf @deboer-tim @mairin 

I have implemented all the above suggestions. Below are screenshots / recordings of the new UI.

I have also decided to put the numbers below the dots, it can encompass 2 digit numbers (I really really doubt that a cluster would have over 100 containers in 1 pod). 

Which makes the UI more uniform in terms of looks. Potentially we could have 1 of every status (total of 9), and numbers underneath it would still fit under the '10 container' limit we've imposed.

https://github.com/containers/podman-desktop/assets/6422176/5a64ba5d-329e-4787-9d54-5c42eca00ca5

![Screenshot 2023-11-06 at 9 49 01 PM](https://github.com/containers/podman-desktop/assets/6422176/d6f48a8d-6577-40f3-a946-11c9cde2876a)

After a preliminary review, I can add more tests for the new files.

benoitf commented:
good work @cdrage 

based on your screenshot, I think the number being displayed below the circle shouldn't move the circles
I mean, circles should always be at the same location on the height of the row (with or without numbers below)


and to my personal preference I think numbers are too big (or the circles are too small) and too bright (but you can ignore my preference :-)

mairin commented:
> Which makes the UI more uniform in terms of looks. Potentially we could have 1 of every status (total of 9), and numbers underneath it would still fit under the '10 container' limit we've imposed.

Hey 👋 That was fast! 😁 I wanna give you my honest opinion Charlie but plz feel free to take or leave it - 

I think the benefit of this approach is you can scale for anything. But a potential issue / downside is that uniformity - just glancing at the list of pods, you no longer get the sense of the comparative "weights" of them... the benefit of the 1:1 icon:container ratio is at a glance you can tell which pod is a biggie and which is small. Now to figure that out, you gotta pause and process the #'s.

I think > 10 containers (and it seems maybe  greater than 3 or 4 really) in a pod is pretty edge case, so you lose the advantage of at l-a-glance relative size to scale to a size you might not ever need to scale to for most users at most times.

The UX approach I take is to optimize for the majority case, so that's why if it were my call I'd say make the relative sizes scannable for < 10... it makes it a little easier for the user to not have to read and thus the overall experience a bit better. But I also don't know for sure and dont have your k8s expertise to know if scanning against relative sizes is important, just a gut feeling.

One way to better enable easier to perceive relative sizing with this approach would be to maybe eventually make the container count a sortable field. You'd probably need to sort by the total # of containers despite the breakdown by type (maybe that would feel weird / awkward to users mechanically but im ot so sure actually.) But then if the relative sizes were important to me, I could sort from most to least containers or vice-versa. 

I hope this helps, it's just what struck me looking this over!

cdrage commented:
> > Which makes the UI more uniform in terms of looks. Potentially we could have 1 of every status (total of 9), and numbers underneath it would still fit under the '10 container' limit we've imposed.
> 
> Hey 👋 That was fast! 😁 I wanna give you my honest opinion Charlie but plz feel free to take or leave it -
> 
> I think the benefit of this approach is you can scale for anything. But a potential issue / downside is that uniformity - just glancing at the list of pods, you no longer get the sense of the comparative "weights" of them... the benefit of the 1:1 icon:container ratio is at a glance you can tell which pod is a biggie and which is small. Now to figure that out, you gotta pause and process the #'s.
> 
> I think > 10 containers (and it seems maybe greater than 3 or 4 really) in a pod is pretty edge case, so you lose the advantage of at l-a-glance relative size to scale to a size you might not ever need to scale to for most users at most times.
> 
> The UX approach I take is to optimize for the majority case, so that's why if it were my call I'd say make the relative sizes scannable for < 10... it makes it a little easier for the user to not have to read and thus the overall experience a bit better. But I also don't know for sure and dont have your k8s expertise to know if scanning against relative sizes is important, just a gut feeling.
> 
> One way to better enable easier to perceive relative sizing with this approach would be to maybe eventually make the container count a sortable field. You'd probably need to sort by the total # of containers despite the breakdown by type (maybe that would feel weird / awkward to users mechanically but im ot so sure actually.) But then if the relative sizes were important to me, I could sort from most to least containers or vice-versa.
> 
> I hope this helps, it's just what struck me looking this over!

Hmmm we are getting there!

Okay so from @benoitf feedback as well as yours, I propose we:

- Make the circles a tad larger since they are quite hard to read to begin with even with less than 10 containers.
- When >10 keep circles still uniform in the middle of the row, but add numbers on bottom (with the current same font size). The larger circles will make it more pleasing to the eye

To be honest, over 10 containers is like you said going to be more of an "astronautical edge case", but we should obviously still have it there for those who do have more than 10 / something odd happened to their pod cluster.

cdrage commented:
Converted to draft as I rebase and make some changes

benoitf commented:
I'm wondering if instead of adding numbers we should not change the shape like to be ellipsis instead of circles (for multiple items) or picking up another form (I won't block the merge of PR anyway, just sharing thoughts)

cdrage commented:
> A non-blocking remark, I'm not sure if it is due to my mild color blindness of if it is a common issue, but I can hardly differentiate between green and red dots when they are not close enough, for example if all dots in one pod are red and all dots in the following pod are green, I hardly differentiate which one is running, which one is stopped

I agree, I added some spacing to fit the numbers in, I'll remove the spacing to make it smaller and more accessible.

cdrage commented:
Ready for review!

Updates:
- > 10 containers the dots are now centered / does not change the view
- Dots are slightly larger now
- No more spacing between dots / they are closer together
- A lot more tests have been added

deboer-tim reviewed (commented):
I really like this, and it's scaling well now. Honestly, the only thing that threw me off was 'why does pink mean stopped'. IMHO stopped/exited should be an outlined circle (no fill) to try to match what we have in the other pages (or the pod itself).

Will need some follow-up to make sure we open issues to use these newly defined colors in other places.

Tests 👍🏼.

cdrage commented:
> I really like this, and it's scaling well now. Honestly, the only thing that threw me off was 'why does pink mean stopped'. IMHO stopped/exited should be an outlined circle (no fill) to try to match what we have in the other pages (or the pod itself).
> 
> Will need some follow-up to make sure we open issues to use these newly defined colors in other places.
> 
> Tests 👍🏼.

I agree, I will add a second set of dots we can add which instead have "outlines" with no fills (example red outline no fill, grey outline no fill, etc.). Which would very beneficial instead of having them confused with a small pink dot vs red, etc.

The second set of dots will allow us to add a grey dot with an outline that's the same as the current icon's we have

cdrage commented:
> I really like this, and it's scaling well now. Honestly, the only thing that threw me off was 'why does pink mean stopped'. IMHO stopped/exited should be an outlined circle (no fill) to try to match what we have in the other pages (or the pod itself).
> 
> Will need some follow-up to make sure we open issues to use these newly defined colors in other places.
> 
> Tests 👍🏼.

Ready for another review! Great suggestion with regards to outline, I've now added it. Having outlines opens up a greater set of possibilities for different colours / representing statuses, even being able to possible integrate this into the cluster UI.

Feel free to go over the colors in tailwind.config.cjs to make sure it's the correct ones!

I've realized that we use the same colour / gray outline for stopped and exited, so I've updated that accordingly in this PR. 

![Screenshot 2023-11-09 at 10 45 32 AM](https://github.com/containers/podman-desktop/assets/6422176/71074719-711e-48ad-82ac-f600282160f8)

deboer-tim reviewed (approved):
Outline works much better and looks consistent!

I still see a slight difference in that the exited status icon is currently using gray-700 for the outline (#aaabac) while the dot outline is showing up as #d1d5db, but we may be into the realm of 'color differences that only Tim would notice' so I am approving and I can investigate some time later.

deboer-tim commented:
(need to fix failing tests though)

cdrage commented:
> Outline works much better and looks consistent!
> 
> I still see a slight difference in that the exited status icon is currently using gray-700 for the outline (#aaabac) while the dot outline is showing up as #d1d5db, but we may be into the realm of 'color differences that only Tim would notice' so I am approving and I can investigate some time later.

I used -700 instead of -300! Fixed it in the config and good to go now!

cdrage commented:
Fixeed the test (was testing against the incorrect class), enabling auto-merge.

mairin commented:
There's still issues here with relying on color, this is how they look in the mockups:

![Screenshot from 2023-11-09 18-49-49](https://github.com/containers/podman-desktop/assets/799683/d7b997fd-da68-4a9d-b17b-1f6acde801ae)

So the orange has an X in it, so you can rely on the shape without the color. The colors from the mockup are:

- green-500 (filled)
- amber-600 (outline x)
- gray-900 (outline)

mairin commented:
gah i missed this will move to an issue
